src/passes/OptimizeInstructions.cpp

```diff
@@ -820,6 +746,103 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          if (select->type == Type::i32) {
+            if (constTrue->value.geti32() == int32_t(1) &&
+                constFalse->value.geti32() == int32_t(0)) {
+              // [i32]->[i32]:   !x ? 1 : 0   ==>   !x
+              if (auto* condition = select->condition->dynCast<Unary>()) {
+                if (condition->isRelational()) {
+                  return condition;
+                }
+              }
+              // [i32]->[i32]:   x <=> y ? 1 : 0   ==>   x <=> y
+              if (auto* condition = select->condition->dynCast<Binary>()) {
+                if (condition->isRelational()) {
+                  return condition;
+                }
+              }
+              // [i32]->[i32]:   expr ? 1 : 0   ==>   expr != 0
+              return Builder(*getModule())
+                .makeBinary(NeInt32, select->condition, constFalse);
+            } else if (constTrue->value.geti32() == int32_t(0) &&
+                       constFalse->value.geti32() == int32_t(1)) {
+              // [i32]->[i32]:   !x ? 0 : 1   ==>   x != 0
+              if (auto* condition = select->condition->dynCast<Unary>()) {
+                if (condition->isRelational()) {
+                  return Builder(*getModule())
+                    .makeBinary(NeInt32, condition->value, constTrue);
```

Comment:
this could add an eqz on the outside, which is shorter?

```
!x ? 0 : 1   ==>   !!x
```

Reply:
Good catch!